src/Microsoft.AspNetCore.Mvc.TagHelpers/TagHelperOutputExtensions.cs

```diff
@@ -154,6 +156,230 @@ public static void MergeAttributes(this TagHelperOutput tagHelperOutput, TagBuil
+        {
+            if (tagHelperOutput == null)
+            {
+                throw new ArgumentNullException(nameof(tagHelperOutput));
+            }
+
+            var encodedSpaceChars = SpaceChars.Where(x => !x.Equals('\u0020')).Select(x => htmlEncoder.Encode(x.ToString())).ToArray();
+
+            if (SpaceChars.Any(classValue.Contains) || encodedSpaceChars.Any(classValue.Contains))
+            {
+                throw new ArgumentException(Resources.ArgumentCannotContainHtmlSpace, nameof(classValue));
+            }
+
+            if (!tagHelperOutput.Attributes.TryGetAttribute("class", out TagHelperAttribute classAttribute))
+            {
+                return;
+            }
+
+            var currentClassValue = ExtractClassValue(classAttribute, htmlEncoder);
+
+            if (string.IsNullOrEmpty(currentClassValue))
+            {
+                return;
+            }
+
+            var encodedClassValue = htmlEncoder.Encode(classValue);
+
+            if (currentClassValue.Equals(encodedClassValue))
+            {
+                tagHelperOutput.Attributes.Remove(tagHelperOutput.Attributes["class"]);
+                return;
+            }
+
+            if (!currentClassValue.Contains(encodedClassValue))
+            {
+                return;
+            }
+
+            var containsSpace = SpaceChars.Any(currentClassValue.Contains);
+
+            var containsEncodedSpace = encodedSpaceChars.Any(currentClassValue.Contains);
+
+            var listOfClasses = new List<string>();
+
+            if (containsSpace && !containsEncodedSpace)
+            {
+                listOfClasses = currentClassValue.Split(SpaceChars).ToList();
+
+                if (!listOfClasses.Contains(encodedClassValue))
+                {
+                    return;
+                }
+            }
+            else if (!containsSpace && containsEncodedSpace)
+            {
+                listOfClasses = currentClassValue.Split(encodedSpaceChars, StringSplitOptions.None).ToList();
+
+                if (!listOfClasses.Contains(encodedClassValue))
+                {
+                    return;
+                }
+            }
+            else if (containsSpace)
+            {
+                listOfClasses = currentClassValue.Split(SpaceChars).ToList();
+                var classesContainingEncodedSpace = listOfClasses.Where(x => encodedSpaceChars.Any(x.Contains));
+                var listOfSubClasses = new List<string>();
+
+                foreach (var classContainingEncodedSpace in classesContainingEncodedSpace)
+                {
+                    var subClasses = classContainingEncodedSpace.Split(encodedSpaceChars, StringSplitOptions.None);
+                    listOfSubClasses.AddRange(subClasses);
+                }
+
+                listOfClasses.RemoveAll(x => encodedSpaceChars.Any(x.Contains));
+                listOfClasses.AddRange(listOfSubClasses);
+
+                if (!listOfClasses.Contains(encodedClassValue))
```

Comment:
You're already declaring `listOfClasses` before the first condition. Suggest moving this repeated check after this block.